src/data/ellpack_page_source.cu

```diff
@@ -3,162 +3,85 @@
     page_size_ = param.gpu_page_size;
   }
 
   monitor_.Init("ellpack_page_source");
-  dh::safe_cuda(cudaSetDevice(device_));
+  dh::safe_cuda(cudaSetDevice(param.gpu_id));
 
   monitor_.StartCuda("Quantiles");
-  common::HistogramCuts hmat;
-  size_t row_stride =
-      common::DeviceSketch(device_, param.max_bin, param.gpu_batch_nrows, dmat, &hmat);
+  common::HistogramCuts cuts;
+  size_t row_stride = common::DeviceSketch(param.gpu_id, param.max_bin,
+                                           param.gpu_batch_nrows, dmat, &cuts);
   monitor_.StopCuda("Quantiles");
 
-  monitor_.StartCuda("CreateEllpackInfo");
-  ellpack_info_ = EllpackInfo(device_, dmat->IsDense(), row_stride, hmat, &ba_);
-  monitor_.StopCuda("CreateEllpackInfo");
-
   monitor_.StartCuda("WriteEllpackPages");
-  WriteEllpackPages(dmat, cache_info);
+  WriteEllpackPages(param.gpu_id, dmat, cuts, cache_info, row_stride);
   monitor_.StopCuda("WriteEllpackPages");
 
-  source_.reset(new ExternalMemoryPrefetcher<EllpackPage>(
-      ParseCacheInfo(cache_info_, kPageType_)));
-}
-
-void EllpackPageSourceImpl::BeforeFirst() {
-  source_.reset(new ExternalMemoryPrefetcher<EllpackPage>(
-      ParseCacheInfo(cache_info_, kPageType_)));
-  source_->BeforeFirst();
-}
-
-bool EllpackPageSourceImpl::Next() {
-  return source_->Next();
-}
-
-EllpackPage& EllpackPageSourceImpl::Value() {
-  EllpackPage& page = source_->Value();
-  page.Impl()->InitDevice(device_, ellpack_info_);
-  return page;
-}
-
-const EllpackPage& EllpackPageSourceImpl::Value() const {
-  EllpackPage& page = source_->Value();
-  page.Impl()->InitDevice(device_, ellpack_info_);
-  return page;
+  external_prefetcher_.reset(
+      new ExternalMemoryPrefetcher<EllpackPage>(cache_info_));
 }
 
 // Compress each CSR page to ELLPACK, and write the accumulated pages to disk.
-void EllpackPageSourceImpl::WriteEllpackPages(DMatrix* dmat, const std::string& cache_info) const {
+void EllpackPageSource::WriteEllpackPages(int device, DMatrix* dmat,
+                                          const common::HistogramCuts& cuts,
+                                          const std::string& cache_info,
+                                          size_t row_stride) const {
   auto cinfo = ParseCacheInfo(cache_info, kPageType_);
   const size_t extra_buffer_capacity = 6;
-  SparsePageWriter<EllpackPage> writer(
-      cinfo.name_shards, cinfo.format_shards, extra_buffer_capacity);
+  SparsePageWriter<EllpackPage> writer(cinfo.name_shards, cinfo.format_shards,
+                                       extra_buffer_capacity);
   std::shared_ptr<EllpackPage> page;
+  SparsePage temp_host_page;
   writer.Alloc(&page);
   auto* impl = page->Impl();
-  impl->matrix.info = ellpack_info_;
-  impl->Clear();
 
-  const MetaInfo& info = dmat->Info();
   size_t bytes_write = 0;
   double tstart = dmlc::GetTime();
   for (const auto& batch : dmat->GetBatches<SparsePage>()) {
-    impl->Push(device_, batch);
+    temp_host_page.Push(batch);
```

Comment:
How expensive is this?

Reply:
This was being done before, except it was happening inside EllpackPage.